rest-service/manager_rest/cluster_status_manager.py

```diff
@@ -0,0 +1,237 @@
+            file_name = '{0}_{1}.json'.format(service_type, node.node_id)
+            status_file_path = path.join(CLUSTER_STATUS_PATH, file_name)
+
+            if not path.exists(status_file_path):
+                self._add_missing_status_file(node, formatted_nodes,
+                                              service_type)
+                continue
+
+            with open(status_file_path, 'r') as status_file:
+                node_status = json.load(status_file)
+
+            if not self._is_report_valid(node_status):
+                self._generate_node_status(node, formatted_nodes,
+                                           status=ServiceStatus.FAIL)
+                continue
+
+            report = node_status['report']
+            self._generate_node_status(node, formatted_nodes,
+                                       report['services'], report['status'])
+
+        is_external = service_nodes[0].to_dict().get('is_external', False)
+        self.cluster_status['services'][service_type] = {
+            'status': self._get_service_status(formatted_nodes),
+            'is_external': is_external,
+            'nodes': formatted_nodes
+        }
+
+    def _generate_node_status(self, node, formatted_nodes, node_services=None,
+                              status=UNINITIALIZED_STATUS):
+        node_dict = node.to_dict()
+        node_name = node_dict.get('name') or node_dict.get('hostname')
+        formatted_nodes[node_name] = {
+            'status': status,
+            'node_id': node_dict.get('node_id'),
+            'services': node_services or {},
+            'version': self.cloudify_version,
+            'public_ip': node_dict.get('public_ip'),
+            'private_ip': (node_dict.get('private_ip') or
+                           node_dict.get('host'))
+        }
+
+    def _add_missing_status_file(self, node, formatted_nodes, service_type):
+        self._generate_node_status(node, formatted_nodes)
+        self.missing_status_files.setdefault(service_type, [])
+        self.missing_status_files[service_type].append(node)
+
+    def _get_entire_cluster_status(self):
+        statuses = [service['status'] for service in
+                    self.cluster_status['services'].values()]
+        if ServiceStatus.FAIL in statuses:
+            return ServiceStatus.FAIL
+        elif ServiceStatus.DEGRADED in statuses:
+            return ServiceStatus.DEGRADED
+        return ServiceStatus.HEALTHY
+
+    def _is_all_in_one(self):
+        return (self.cluster_structure[ServiceType.DB][0].node_id ==
+                self.cluster_structure[ServiceType.BROKER][0].node_id ==
+                self.cluster_structure[ServiceType.MANAGER][0].node_id)
+
+    def _get_service_status(self, formatted_nodes):
+        # A set of the service node statuses
+        service_statuses = {node['status'] for name, node in
+                            formatted_nodes.items()}
+
+        # If there is only one type of status it is it, else Degraded
```

Comment:
this comment doesn't really help, because it just says the same thing that the code says

instead if you need a comment here, it should talk about why is it degraded if there's not exactly 1 node: i'd think having 3 nodes isn't degraded? we want to have 3 db nodes, 3 rabbit nodes...

Reply:
It checks the different statuses type, not matter how many node there are.
If all the node statuses are the same = OK or Fail this will be the service status.
Else it means some Fail and some OK and the service status is Degraded.

Reply:
Only the content of the comment was not clear